src/main/java/org/folio/service/financedata/FinanceDataService.java

```diff
@@ -45,18 +53,93 @@ public FinanceDataService(FundService fundService, BudgetService budgetService,
-      })
+      .withTrans(conn -> fiscalYearService.getFiscalYearById(getFiscalYearId(entity), conn)
+        .compose(fiscalYear -> createBudgets(entity, fiscalYear, conn, okapiHeaders)
+          .compose(v -> updateFundAndBudget(entity, conn))
+          .compose(v -> processAllocationTransaction(entity, fiscalYear, conn, okapiHeaders))))
       .onSuccess(v -> logger.info("Successfully updated finance data"))
-      .onFailure(e -> logger.error("Failed to update finance data", e))
+      .onFailure(e -> logger.error("Failed to update finance data", e));
+  }
+
+  private Future<Void> createBudgets(FyFinanceDataCollection entity, FiscalYear fiscalYear, DBConn conn,
+      Map<String, String> okapiHeaders) {
+    if (fiscalYearIsPast(fiscalYear)) {
+      return succeededFuture();
+    }
+    List<FyFinanceData> dataList = entity.getFyFinanceData().stream()
+      .filter(data -> data.getBudgetId() == null && (data.getBudgetStatus() != null ||
+        requireNonNullElse(data.getBudgetAllocationChange(), 0.0) > 0))
+      .toList();
+    if (dataList.isEmpty()) {
+      return succeededFuture();
+    }
+    List<Budget> budgets = dataList.stream()
+      .map(data -> {
+        String budgetId = UUID.randomUUID().toString();
+        Budget.BudgetStatus budgetStatus = newBudgetStatus(fiscalYear);
+        data.setBudgetId(budgetId);
+        if (data.getBudgetStatus() == null) {
+          data.setBudgetStatus(FyFinanceData.BudgetStatus.fromValue(budgetStatus.value()));
+        }
+        Budget budget = new Budget()
+          .withId(budgetId)
+          .withName(requireNonNullElse(data.getBudgetName(), data.getFundCode() + '-' + data.getFiscalYearCode()))
+          .withBudgetStatus(budgetStatus)
+          .withAllocated(0.)
+          .withFundId(data.getFundId())
+          .withFiscalYearId(data.getFiscalYearId());
+        populateBudgetMetadata(budget, okapiHeaders);
+        return budget;
+      })
+      .toList();
+    if (budgets.isEmpty()) {
+      return succeededFuture();
+    }
+    return budgetService.createBatchBudgets(budgets, conn);
+  }
+
+  private boolean fiscalYearIsPast(FiscalYear fiscalYear) {
+    Date now = new Date();
+    return fiscalYear.getPeriodEnd().before(now);
+  }
+
+  private Budget.BudgetStatus newBudgetStatus(FiscalYear fiscalYear) {
+    Date now = new Date();
+    Budget.BudgetStatus budgetStatus;
+    if (fiscalYear.getPeriodStart().before(now)) {
+      budgetStatus = Budget.BudgetStatus.ACTIVE;
+    } else {
+      budgetStatus = Budget.BudgetStatus.PLANNED;
+    }
+    return budgetStatus;
+  }
+
+  private void populateBudgetMetadata(Budget budget, Map<String, String> okapiHeaders) {
+    String userId = okapiHeaders.get(XOkapiHeaders.USER_ID);
+    if (userId == null) {
+      try {
+        userId = (new OkapiToken(okapiHeaders.get(XOkapiHeaders.TOKEN))).getUserIdWithoutValidation();
+      } catch (Exception ignored) {
+        // could not find user id - ignoring
```

Comment:
we can add log warn

Reply:
This method is also used for batch transactions, and is a copy of code from Okapi. I think an exception here is a normal case, and we should not print a warning (Okapi does not).